<a href="https://colab.research.google.com/github/melchiadesblanco/CID_UX4DS/blob/main/CASE03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introdução
Case 3 - Avaliação Modelo NBO

#### O objetivo deste trabalho é: 

Contar uma história que mostre se a recomendação do carrossel faz ou não sentido com base no histórico de compra do cliente.

A idéia é aplicar princípios de validação da metodologia CRISP.

Se você fosse um coordenador de DS, você deixaria este modelo ser implementado em produção?

NBO -> "Next Best Offer"
No dataset atual, é feito uma sugestão de 5 produtos e também traz o score de cada sugestão que é a probabilidade do consumidor comprar aquele produto.

#### Resumo Aula

- Aula 1: Criação dos grupos e introdução aos temas
- Aula 2: Fazer as análises exploratórias básicas

# Anotações do Professor
- 12M faz sentido olhando para 6M?
- Valores errados? Nulos? Zero?
- Outliers
- Distribuição dos produtos vendidos faz sentido?

# Problemas Encontrados
1. As recomendações do carrosel utilizam um range de sugestões para cada NBO, o que pode tornar alguns NBO mal utilizados.

- NBO 1 -> Produto 1  ao 20
- NBO 2 -> Produto 21 ao 33
- NBO 3 -> Produto 34 ao 50
- NBO 4 -> Produto 51 ao 66
- NBO 5 -> Produto 67 ao 80

2. Foi observado que existe uma grande quantidade de recomendações unicas de NBOs, ou seja, que não são recomendadas para outros consumidores com os mesmos valores de NBO 1-5.
- Isto pode significar que o modelo pode estar com overfiting

## Limpeza do Dataset

In [7]:
import numpy as np
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [17]:
#Rodar no Jupyter Local
# df_hist = pd.read_csv('./data/CASE_03_Compras_Hist.csv', sep=';')
# df_carrossel = pd.read_csv('./data/CASE_03_NBO_Result.csv', sep=';')

#Rodar no Google Colab
df_hist = pd.read_csv('/content/CASE_03_Compras_Hist.csv', sep=';')
df_carrossel = pd.read_csv('/content/CASE_03_NBO_Result.csv', sep=';')

In [15]:
df_hist.dtypes

Consumidor            int64
Produto               int64
Quantidade Acc 12M    int64
Quantidade Acc 6M     int64
dtype: object

In [18]:
df_carrossel['Score NBO 1'] = df_carrossel['Score NBO 1'].str.replace(',', '.').astype(float)
df_carrossel['Score NBO 2'] = df_carrossel['Score NBO 2'].str.replace(',', '.').astype(float)
df_carrossel['Score NBO 3'] = df_carrossel['Score NBO 3'].str.replace(',', '.').astype(float)
df_carrossel['Score NBO 4'] = df_carrossel['Score NBO 4'].str.replace(',', '.').astype(float)
df_carrossel['Score NBO 5'] = df_carrossel['Score NBO 5'].str.replace(',', '.').astype(float)

In [19]:
df_carrossel.dtypes

Consumidor       int64
NBO 1            int64
NBO 2            int64
NBO 3            int64
NBO 4            int64
NBO 5            int64
Score NBO 1    float64
Score NBO 2    float64
Score NBO 3    float64
Score NBO 4    float64
Score NBO 5    float64
dtype: object

In [21]:
#Valores Nulos
df_hist[df_hist.isna().any(axis=1)]

Empty DataFrame
Columns: [Consumidor, Produto, Quantidade Acc 12M, Quantidade Acc 6M]
Index: []

In [22]:
df_carrossel[df_carrossel.isna().any(axis=1)]

Empty DataFrame
Columns: [Consumidor, NBO 1, NBO 2, NBO 3, NBO 4, NBO 5, Score NBO 1, Score NBO 2, Score NBO 3, Score NBO 4, Score NBO 5]
Index: []

In [23]:
#Valores inconsistentes
df_hist[df_hist['Quantidade Acc 12M'] < df_hist['Quantidade Acc 6M']]

Empty DataFrame
Columns: [Consumidor, Produto, Quantidade Acc 12M, Quantidade Acc 6M]
Index: []

## Análise Exploratória

In [24]:
#Analise Inicial - Histórico de Compras
df_hist.describe()

Consumidor     Produto  Quantidade Acc 12M  Quantidade Acc 6M
count 1048488.000 1048488.000         1048488.000        1048488.000
mean     9362.000      40.493               2.500              0.625
std      5404.867      23.082               1.708              1.099
min         1.000       1.000               0.000              0.000
25%      4681.000      21.000               1.000              0.000
50%      9362.000      41.000               2.000              0.000
75%     14043.000      60.000               4.000              1.000
max     18723.000      80.000               5.000              5.000

In [25]:
df_hist.sample(n=5)

Consumidor  Produto  Quantidade Acc 12M  Quantidade Acc 6M
637832       11390       16                   0                  0
705070       12591        8                   3                  0
314816        5622       69                   4                  0
287177        5129       11                   4                  0
608256       10862       32                   0                  0

In [26]:
#Analise Inicial - Histórico de Recomendações
df_carrossel.describe()

Consumidor     NBO 1     NBO 2  ...  Score NBO 3  Score NBO 4  Score NBO 5
count   18723.000 18723.000 18723.000  ...    18723.000    18723.000    18723.000
mean     9362.000    10.574    27.017  ...        0.745        0.746        0.745
std      5405.009     5.778     3.741  ...        0.138        0.138        0.139
min         1.000     1.000    21.000  ...        0.510        0.510        0.510
25%      4681.500     6.000    24.000  ...        0.630        0.630        0.620
50%      9362.000    11.000    27.000  ...        0.750        0.750        0.740
75%     14042.500    16.000    30.000  ...        0.860        0.870        0.870
max     18723.000    20.000    33.000  ...        0.980        0.980        0.980

[8 rows x 11 columns]

In [27]:
df_carrossel.sample(n=5)

Consumidor  NBO 1  NBO 2  ...  Score NBO 3  Score NBO 4  Score NBO 5
13848       13849     11     29  ...        0.830        0.850        0.570
9396         9397      6     32  ...        0.560        0.970        0.510
14820       14821     13     28  ...        0.570        0.930        0.680
8431         8432     14     25  ...        0.720        0.670        0.910
8488         8489      9     31  ...        0.950        0.720        0.570

[5 rows x 11 columns]

In [28]:
#TOP 5 Produtos vendidos na base histórica
df_hist_top5 = df_hist['Produto'].value_counts().sort_values(ascending=False).head(5)
df_hist_top5

56    13444
19    13387
3     13360
14    13291
76    13284
Name: Produto, dtype: int64

In [29]:
#LOWEST 5 Produtos vendidos na base histórica
df_hist_lowest5 = df_hist['Produto'].value_counts().sort_values(ascending=True).head(5)
df_hist_lowest5

13    12845
47    12847
12    12873
26    12885
30    12907
Name: Produto, dtype: int64

In [30]:
#TOP 5 Produtos Mais Sugeridos no carrossel
a = df_carossel['NBO 1']
b = df_carossel['NBO 2']
c = df_carossel['NBO 3']
d = df_carossel['NBO 4']
e = df_carossel['NBO 5']

df_carrossel_top5 = pd.concat([a, b, c, d, e])
df_carrossel_top5 = df_carrossel_top5.value_counts().sort_values(ascending=False).head(5)
df_carrossel_top5

31    1476
27    1471
32    1463
22    1463
24    1460
dtype: int64

In [31]:
#LOWEST 5 Produtos Menos Sugeridos no carrossel
df_carrossel_lowest5 = pd.concat([a, b, c, d, e])
df_carrossel_lowest5 = df_carrossel_lowest5.value_counts().sort_values(ascending=True).head(5)
df_carrossel_lowest5

14    875
3     891
8     899
7     904
2     907
dtype: int64

In [32]:
#TOP 5 Consumidores (com base no numero de compras na base histórica) 6M
df_hist_top5con6M = df_hist[df_hist['Quantidade Acc 6M'] > 0]
df_hist_top5con6M = df_hist_top5con6M['Consumidor'].value_counts().sort_values(ascending=False).head(5)
df_hist_top5con6M

# Quantidade Acc 12M 	Quantidade Acc 6M

1008     33
4265     31
11142    31
12681    31
11653    31
Name: Consumidor, dtype: int64

In [33]:
#TOP 5 Consumidores (com base no numero de compras na base histórica) 12M
df_hist_top5con12M = df_hist[df_hist['Quantidade Acc 12M'] > 0]
df_hist_top5con12M = df_hist_top5con12M['Consumidor'].value_counts().sort_values(ascending=False).head(5)
df_hist_top5con12M

9406     56
4983     55
15623    55
13090    55
5686     55
Name: Consumidor, dtype: int64

In [34]:
#Trazer o NBO do TOP 5 Consumidores 6M
df_hist_top5con6M = pd.merge(df_hist_top5con6M, df_carrossel, how='inner', on=["Consumidor", "Consumidor"])
df_hist_top5con6M

Consumidor  NBO 1  NBO 2  ...  Score NBO 3  Score NBO 4  Score NBO 5
0          33     10     22  ...        0.840        0.880        0.560
1          31     11     23  ...        0.880        0.750        0.640
2          31     11     23  ...        0.880        0.750        0.640
3          31     11     23  ...        0.880        0.750        0.640
4          31     11     23  ...        0.880        0.750        0.640

[5 rows x 11 columns]

In [35]:
#Trazer o NBO do TOP 5 Consumidores 12M
df_hist_top5con12M = pd.merge(df_hist_top5con12M, df_carrossel, how='inner', on=["Consumidor", "Consumidor"])
df_hist_top5con12M

Consumidor  NBO 1  NBO 2  ...  Score NBO 3  Score NBO 4  Score NBO 5
0          56      9     31  ...        0.870        0.530        0.930
1          55      4     28  ...        0.900        0.950        0.540
2          55      4     28  ...        0.900        0.950        0.540
3          55      4     28  ...        0.900        0.950        0.540
4          55      4     28  ...        0.900        0.950        0.540

[5 rows x 11 columns]

# REMARK: Qual a lógica por traz da recomendação dos 5 produtos do carrossel?
Cada NBO [1 ao 5] pode sugerir apenas um range de produtos como é demonstrado a seguir

=> Isto é ruim pois caso o cliente tenha um historico de compras somente para produtos da faixa (67 ao 80), os outros NBOs ficariam mal utilizados

In [36]:
df_carrossel['NBO 1'].describe()   # Produto 1  ao 20

count   18723.000
mean       10.574
std         5.778
min         1.000
25%         6.000
50%        11.000
75%        16.000
max        20.000
Name: NBO 1, dtype: float64

In [37]:
df_carrossel['NBO 2'].describe()    # Produto 21 ao 33

count   18723.000
mean       27.017
std         3.741
min        21.000
25%        24.000
50%        27.000
75%        30.000
max        33.000
Name: NBO 2, dtype: float64

In [38]:
df_carrossel['NBO 3'].describe()    # Produto 34 ao 50

count   18723.000
mean       42.057
std         4.918
min        34.000
25%        38.000
50%        42.000
75%        46.000
max        50.000
Name: NBO 3, dtype: float64

In [39]:
df_carrossel['NBO 4'].describe()    # Produto 51 ao 66

count   18723.000
mean       58.506
std         4.625
min        51.000
25%        55.000
50%        58.000
75%        63.000
max        66.000
Name: NBO 4, dtype: float64

In [40]:
df_carrossel['NBO 5'].describe()    # Produto 67 ao 80

count   18723.000
mean       73.476
std         4.025
min        67.000
25%        70.000
50%        73.000
75%        77.000
max        80.000
Name: NBO 5, dtype: float64

# REMARK: Quantas variações de recomendações foram feitas?
==> 18562 variações de um total de 18723 => 99.14 % de recomendações únicas

In [41]:
def left(s, amount):
    return s[:amount]

def right(s, amount):
    return s[-amount:]

def mid(s, offset, amount):
    return s[offset:offset+amount]


df_carrossel['ConjuntoID'] = 99*(100**5) + (df_carrossel['NBO 1']*(100**4)) + (df_carrossel['NBO 2']*(100**3)) + (df_carrossel['NBO 3']*(100**2)) + (df_carrossel['NBO 4']*(100**1)) + (df_carrossel['NBO 5']*(100**0))
df_carrossel['ConjuntoID'].value_counts().sort_values(ascending=False)


990425376570    3
991624426067    2
990722505178    2
990721505274    2
990124375374    2
               ..
990821436679    1
990222505973    1
992033506378    1
991732445178    1
990233435473    1
Name: ConjuntoID, Length: 18562, dtype: int64

#REMARK: Consumidores que nunca compraram?

In [70]:
df_hist_nonbuyers = df_hist.groupby(['Consumidor'], as_index=False)['Quantidade Acc 12M'].sum()
df_hist_nonbuyers.columns = ['Consumidor', 'Total Acc 12M']
df_hist_nonbuyers[df_hist_nonbuyers['Total Acc 12M']==0 ]

Empty DataFrame
Columns: [Consumidor, Total Acc 12M]
Index: []

#REMARK: Produtos que nunca foram vendidos?

In [72]:
df_hist_neversold = df_hist.groupby(['Produto'], as_index=False)['Quantidade Acc 12M'].sum()
df_hist_neversold.columns = ['Produto', 'Total Acc 12M']
df_hist_neversold[df_hist_neversold['Total Acc 12M']==0 ]

Empty DataFrame
Columns: [Produto, Total Acc 12M]
Index: []